server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java

```diff
@@ -0,0 +1,213 @@
+ */
+
+package org.elasticsearch.repositories;
+
+import org.elasticsearch.common.Nullable;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public final class ShardGenerations {
+
+    public static final ShardGenerations EMPTY = new ShardGenerations(Collections.emptyMap());
+
+    /**
+     * Special generation that signifies that a shard is new and the repository does not yet contain a valid
+     * {@link org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshots} blob for it.
+     */
+    public static final String NEW_SHARD_GEN = "_new";
+
+    /**
+     * Special generation that signifies that the shard has been deleted from the repository.
+     * This generation is only used during computations. It should never be written to disk.
+     */
+    public static final String DELETED_SHARD_GEN = "_deleted";
+
+    private final Map<IndexId, List<String>> shardGenerations;
+
+    private ShardGenerations(Map<IndexId, List<String>> shardGenerations) {
+        this.shardGenerations = shardGenerations;
+    }
+
+    /**
+     * Returns all indices for which shard generations are tracked.
+     *
+     * @return indices for which shard generations are tracked
+     */
+    public Collection<IndexId> indices() {
+        return Collections.unmodifiableSet(shardGenerations.keySet());
+    }
+
+    /**
+     * Computes the obsolete shard index generations that can be deleted once this instance was written to the repository.
+     * Note: This method should only be used when finalizing a snapshot and we can safely assume that data has only been added but not
+     *       removed from shard paths.
+     *
+     * @param previous Previous {@code ShardGenerations}
+     * @return Map of obsolete shard index generations in indices that are still tracked by this instance
+     */
+    public Map<IndexId, Map<Integer, String>> obsoleteShardGenerations(ShardGenerations previous) {
+        final Map<IndexId, Map<Integer, String>> result = new HashMap<>();
+        previous.shardGenerations.forEach(((indexId, oldGens) -> {
+            final List<String> updatedGenerations = shardGenerations.get(indexId);
+            final Map<Integer, String> obsoleteShardIndices = new HashMap<>();
+            assert updatedGenerations != null
+                : "Index [" + indexId + "] present in previous shard generations, but missing from updated generations";
+            for (int i = 0; i < Math.min(oldGens.size(), updatedGenerations.size()); i++) {
+                final String oldGeneration = oldGens.get(i);
+                final String updatedGeneration = updatedGenerations.get(i);
+                if (updatedGeneration != null && oldGeneration != null && oldGeneration.equals(updatedGeneration) == false) {
```

Comment:
can you add a comment here, saying why we only override info for cases where we have oldGeneration as well as updatedGeneration?

Reply:
Done